### PR TITLE
Fix #417 preserve location of trailing loud comments

### LINKED ISSUE #417: Inline comments should be kept inline
This is a "repeat" of the ticket made at sass/sass#235 which presented the problem of code like:
```
.myselector {
  display: none; /* javascript will fix this */
}
```
compiling to:
```
.myselector {
  display: none;
  /* javascript will fix this */
}
```
That ticket was moved to ruby-sass and closed as on ice since it was deemed to be specific to Ruby Sass's deprecated implementation. However, I'm seeing it happen for me despite using Dart Sass. I use `--style expanded` for what it's worth.

My apologies if this is somehow being caused by interference with an old and probably only partially working ruby (when I run `sass --help` I am seeing references to Dart, e.g. `    --version                  Print the version of Dart Sass.`, though don't know if there's still possible room for interference, or how I might resolve that). Thanks for any possible work on this, or letting me know if there is simply some flag or style setting whose documentation I missed.

### PR DESCRIPTION
**Pull request by nickbehrens**
Fix #417 by preserving the location of trailing loud comments.

Loud comment AST nodes now have the notion of whether they're "trailing".  During parsing, we look for loud comments at the end of lines and if found, tag them as trailing.  This information is then plumbed through and made available during serialization.  

During serialization, where we would previously unconditionally insert newlines between nodes, we now check to see if the next node is a trailing loud comment.  If so, we instead write out a space followed by the trailing loud comment.

Ran the following locally, all tests passing:
$ pub run grinder && dartanalyzer lib test && pub run test -x node

See https://github.com/sass/sass-spec/pull/1485

### CODE REVIEW

File: lib/src/visitor/serialize.dart [hunk 1]
```diff
@@ -1085,23 +1090,41 @@ class _SerializeVisitor
...
+      _writeOptionalSpace();
+      indentNextChild = false;
+    }
+
     CssNode previous;
-    _indent(() {
-      for (var i = 0; i < children.length; i++) {
-        var child = children[i];
-        if (_isInvisible(child)) continue;
+    for (var i = 0; i < children.length; i++) {
```
Comment (nex3) [inline lib/src/visitor/serialize.dart]:
```suggestion
    for (var child in children) {
```

I don't know why this was using an int-style for loop in the first place.

File: lib/src/visitor/serialize.dart [hunk 2]
```diff
@@ -1085,23 +1090,41 @@ class _SerializeVisitor
...
+          indentNextChild = false;
         }
-        previous = child;
+        if (previous.isGroupEnd) _writeLineFeed();
+      }
+      previous = child;
 
-        child.accept(this);
+      if (indentNextChild) {
+        _indentation++;
```
Comment (nex3) [inline lib/src/visitor/serialize.dart]: I'm not sure I understand this indentation logic. It looks like if `indentNextChild` is false, it will print the child at one level of indentation lower than it would otherwise, but don't we want *zero* indentation for trailing comments?

I wonder if it makes more sense to completely special-case these comments, and write something more like:

```dart
if (_isTrailingComment(child)) {
  _writeOptionalSpace();
  _withoutIndentation(() => child.accept(this));
  continue;
}
```

For the post-bracket comments you'd want to do instead something like:

```dart
if (children.isNotEmpty && _isTrailingComment(children.first)) {
  _writeOptionalSpace();
  _withoutIndentation(() => children.first.accept(this));
  children = children.skip(1);
}
```

File: test/output_test.dart [hunk 3]
```diff
@@ -59,4 +59,34 @@ void main() {
       });
     });
   });
+
+  // Regression test for sass/dart-sass#417.
+  group("preserve trailing loud comments", () {
+    // No need for "in Sass" cases as it's not possible to have
+    // trailing loud comments in the Sass syntax.
+    group("in SCSS", () {
```
Comment (nex3) [inline test/output_test.dart]: I'd just fold this into the previous group name to save on some indentation.

File: test/output_test.dart [hunk 4]
```diff
@@ -59,4 +59,34 @@ void main() {
       });
     });
   });
+
+  // Regression test for sass/dart-sass#417.
```
Comment (nex3) [inline test/output_test.dart]: I wouldn't call this a regression test, since it's really more of a feature.

File: test/output_test.dart [hunk 5]
```diff
@@ -59,4 +59,34 @@ void main() {
...
+    group("in SCSS", () {
+      test("after open block", () {
+        expect(compileString("""
+selector { /* please don't move me */
+  name: value;
+}"""), equals("""
+selector { /* please don't move me */
+  name: value;
+}"""));
+      });
```
Comment (nex3) [inline test/output_test.dart]: Style nit: add empty lines between each test case.

File: test/output_test.dart [hunk 6]
```diff
@@ -59,4 +59,34 @@ void main() {
...
+      test("after declaration", () {
+        expect(compileString("""
+selector {
+  name: value; /* please don't move me */
```
Comment (nex3) [inline test/output_test.dart]: I'd test this both as the last declaration in a block and as a middle declaration. Same with at-rules.

File: test/output_test.dart [hunk 7]
```diff
@@ -59,4 +59,34 @@ void main() {
       });
     });
   });
+
+  // Regression test for sass/dart-sass#417.
+  group("preserve trailing loud comments", () {
+    // No need for "in Sass" cases as it's not possible to have
+    // trailing loud comments in the Sass syntax.
+    group("in SCSS", () {
+      test("after open block", () {
```
Comment (nex3) [inline test/output_test.dart]: Also test what happens if the comment is the only child of the block.

File: lib/src/visitor/serialize.dart (hunk 2 above)
Comment (nickbehrens) [reply lib/src/visitor/serialize.dart]: I tried to incorporate your feedback here. PTAL.

File: test/output_test.dart (hunk 3 above)
Comment (nickbehrens) [reply test/output_test.dart]: done

File: test/output_test.dart (hunk 4 above)
Comment (nickbehrens) [reply test/output_test.dart]: yep, copy-paste from above, removed "regression"

File: test/output_test.dart (hunk 6 above)
Comment (nickbehrens) [reply test/output_test.dart]: done

File: test/output_test.dart (hunk 7 above)
Comment (nickbehrens) [reply test/output_test.dart]: done

File: test/output_test.dart (hunk 5 above)
Comment (nickbehrens) [reply test/output_test.dart]: done

**Comment by nickbehrens** [comment]
> Apologies for not realizing this in advance, but I think you can actually do this without touching the parser at all. A comment is trailing iff it starts on the same line as the previous (or parent) statement, right? So you can just compare their `span.start.line` values in the serializer to tell if the comment should be treated as "trailing".

No worries.  I've reverted the parser changes and instead am now using the span line numbers to look for trailing comments.

File: lib/src/parse/parser.dart [hunk 8]
```diff
@@ -79,11 +79,17 @@ class Parser {
...
-  void whitespaceWithoutComments() {
+  bool whitespaceWithoutComments() {
+    bool sawAnyNewlines = false;
     while (!scanner.isDone && isWhitespace(scanner.peekChar())) {
-      scanner.readChar();
+      if (isNewline(scanner.readChar())) {
+        sawAnyNewlines = true;
+      }
     }
+    return sawAnyNewlines;
```
[the remark this answers is not in the document]
Comment (nickbehrens) [reply lib/src/parse/parser.dart]: thanks for catching, reverting

File: lib/src/visitor/serialize.dart [hunk 9]
```diff
@@ -1078,30 +1080,47 @@ class _SerializeVisitor
       _for(value, () => _buffer.write(value.value));
 
   /// Emits [children] in a block.
-  void _visitChildren(List<CssNode> children) {
+  void _visitChildren(Iterable<CssNode> children, CssNode parent) {
```
[the remark this answers is not in the document]
Comment (nickbehrens) [reply lib/src/visitor/serialize.dart]: ISTR there being a compiler error that wouldn't go away until I switched away from List.  But I'm not able to repro by reverting locally so I'll drop that change.

File: test/output_test.dart [hunk 10]
```diff
@@ -59,4 +59,73 @@ void main() {
...
+      expect(compileString("""
+selector {
+  name: value;
+} /* please don't move me */"""), equals("""
+selector {
+  name: value;
+} /* please don't move me */"""));
+    });
+
+    test("only content in block", () {
```
[the remark this answers is not in the document]
Comment (nickbehrens) [reply test/output_test.dart]: adding test case for this

File: lib/src/visitor/serialize.dart [hunk 11]
```diff
@@ -1115,6 +1134,22 @@ class _SerializeVisitor
...
+  /// Whether [node] represents a trailing comment when it appears after
+  /// [previous] in a sequence of nodes.  If there are no nodes before [node],
+  /// then [previous] should be the parent node.
+  bool _isTrailingComment(CssNode node, CssNode previous) {
+    if (node is CssComment) {
+      // While the last condition (comment start line == previous end line)
+      // seems like it should be sufficient, it doesn't handle the case where
+      // "previous" is a parent node and the trailing comment appears right
+      // after a block open lbrace.
+      return node.span.start.line == previous.span.start.line
```
[the remark this answers is not in the document]
Comment (nickbehrens) [reply lib/src/visitor/serialize.dart]: adding test case for this

[the remark this answers is not in the document]
Comment (nickbehrens) [reply lib/src/visitor/serialize.dart]: I updated the PR to do as you suggested but I had trouble figuring out how to implement the beforeChildren getter in CssParentNode (see the [TODO in CssStylesheet](https://github.com/sass/dart-sass/compare/master...nickbehrens:fix-417#diff-493a0350d1b87914576cbfa5c665bb64)).

Putting the getter impl just in CssParentNode didn't seem to work for classes that merely implement (vs extend) CssParentNode (like ModifiableCssParentNode).

File: test/output_test.dart (hunk 10 above)
[the remark this answers is not in the document]
Comment (nickbehrens) [reply test/output_test.dart]: FYI the trailing comment's location is preserved but a new newline is added before the closing brace (after the trailing comment).  I added a comment about where this happens in the serializer but wasn't sure it was worth the effort to try to change existing behavior: https://github.com/sass/dart-sass/compare/master...nickbehrens:fix-417#diff-872a211685a7abf16c5f7d36c967296cR1130

File: lib/src/visitor/serialize.dart [hunk 12]
```diff
@@ -1078,30 +1080,47 @@ class _SerializeVisitor
...
 
-    _writeLineFeed();
+    bool indentNextChild;
+    if (children.isNotEmpty && _isTrailingComment(children.first, parent)) {
+      _writeOptionalSpace();
+      indentNextChild = false;
+    } else {
+      _writeLineFeed();
+      indentNextChild = true;
+    }
```
[the remark this answers is not in the document]
Comment (nickbehrens) [reply lib/src/visitor/serialize.dart]: No, I don't think I forgot to push anything.  The [newer logic](https://github.com/sass/dart-sass/commit/d937f878dabf46abe490c60b3fd4e3671a6ce299#diff-872a211685a7abf16c5f7d36c967296cR1091) should (a) avoid indenting at all if no indentation is necessary (vs previously just indenting one level less) and (b) special-case trailing comments.

I did also try using your suggested code but I wasn't able to make it easily work with the rest of the existing logic in the function (esp the bits around handling semicolons and newlines that follow those).  Unfortunately it's been a little while since I tried so I don't remember all the details.

File: lib/src/visitor/serialize.dart (hunk 11 above)
[the remark this answers is not in the document]
Comment (nickbehrens) [reply lib/src/visitor/serialize.dart]: Looking at this again a little later, I realized I don't really need a duplicate impl in CssStylesheet (as there's no beforeChildren span that makes sense there).

File: lib/src/visitor/serialize.dart (hunk 12 above)
[the remark this answers is not in the document]
Comment (nickbehrens) [reply lib/src/visitor/serialize.dart]: I took another pass at simplifying _visitChildren and managed to get rid of the indentNextChild var.  I'm guessing this is more similar to what you were originally thinking (?).  PTAL.

File: test/output_test.dart (hunk 10 above)
[the remark this answers is not in the document]
Comment (nickbehrens) [reply test/output_test.dart]: When simplifying _visitChildren, I found a relatively easy way to "fix" this but unfortunately it breaks a bunch of tests that expect the current behavior (e.g. https://github.com/sass/dart-sass/blob/master/test/dart_api_test.dart#L25 loads in SCSS like "a { b: c; }" that lacks any newlines but then expects it to be formatted with newlines after compilation: "a {\n  b: c;\n}").

I'm thinking we should leave this as is.

[the remark this answers is not in the document]
Comment (nex3) [reply test/output_test.dart]: I've pushed a change that gets rid of the trailing newline only for comments, by tracking the *two* previous nodes and checking them to see if the last node was a trailing comment of the parent node.

File: lib/src/visitor/serialize.dart [hunk 13]
```diff
@@ -1078,43 +1080,59 @@ class _SerializeVisitor
...
+  /// Whether [node] represents a trailing comment when it appears after
+  /// [previous] in a sequence of nodes being serialized.  Note [previous]
+  /// could either be (1) a sibling of [node] or (2) the parent of [node], with
+  /// [node] being the first visible child.
+  bool _isTrailingComment(CssNode node, CssNode previous) {
+    if (node is! CssComment) return false;
+    FileSpan previousSpan = previous.span;
+    if (previous is CssParentNode && previous.children.contains(node)) {
+      var lbracePattern = String.fromCharCode($lbrace);
+      var endOffset = math.max(0, previousSpan.text.indexOf(lbracePattern));
```
Comment (nex3) [inline lib/src/visitor/serialize.dart]: It's an edge case, but this `indexOf()` check could fail if a selector or at-rule contains curly braces (for example, `[href*="{"]` is a valid selector), or even if it contains a comment that was removed from the AST (such as `a /*{*/ b { ... }`). It would be safer to look backwards from `node.span.start`.

File: test/output_test.dart (hunk 10 above)
[the remark this answers is not in the document]
Comment (nickbehrens) [reply test/output_test.dart]: ok, thanks!

File: lib/src/visitor/serialize.dart (hunk 13 above)
Comment (nickbehrens) [reply lib/src/visitor/serialize.dart]: I added a test case for this, confirmed it failed, and then fixed it (by doing what you suggested).  I'll push those as soon as all the other tests finish running.

Comment (nickbehrens) [reply lib/src/visitor/serialize.dart]: committed in https://github.com/nickbehrens/dart-sass/commit/45afc0a31b5115487fa816120aff32019890d244

File: test/output_test.dart [hunk 14]
```diff
@@ -145,5 +145,15 @@ selector {
...
     });
+
+    // The trailing comment detection logic looks for left braces to detect
+    // whether a comment is on the same line as its parent node.  This test
+    // checks to make sure it isn't confused by syntax that uses braces for
+    // things other than starting child blocks.
+    test("selector contains left brace", () {
+      expect(compileString("""selector[href*=\"{\"]
+{ /* please don't move me */ }"""),
+          equals("""selector[href*=\"{\"] { /* please don't move me */ }"""));
```
Comment (nex3) [inline test/output_test.dart]:
```suggestion
          equals('selector[href*="{"] { /* please don't move me */ }'));
```

File: lib/src/visitor/serialize.dart [hunk 15]
```diff
@@ -1138,7 +1138,13 @@ class _SerializeVisitor
 
     var previousSpan = previous.span;
     if (previous is CssParentNode && previous.children.contains(node)) {
-      var endOffset = math.max(0, previousSpan.text.indexOf("{"));
+      // Walk back from just before the current node starts looking for the
+      // parent's left brace (to open the child block).  This is safer than
+      // a simple forward search of the previousSpan.text as that might contain
+      // other left braces.
+      var endOffset =
+          previousSpan.text.lastIndexOf("{", node.span.start.offset - 1);
```
Comment (nex3) [inline lib/src/visitor/serialize.dart]: I don't think this logic is quite right... `node.span.start.offset` is an index into the entire source file, while `previousSpan.text`, just covers the span itself. I think you'll need to access [`node.span.file.getText()`](https://pub.dev/documentation/source_span/latest/source_span/SourceFile/getText.html).

Comment (nickbehrens) [reply lib/src/visitor/serialize.dart]: ah, I thought it was an offset relative to the beginning of the region covered by the span.  But I see now the FileLocation APIs are all absolute (which makes sense).  

I made the test case a little richer to try to catch the issue and then updated isTrailingComment to convert the absolute offset to relative to previousSpan's start offset: https://github.com/nickbehrens/dart-sass/commit/35f51df40734af3653e8f049d1915cbf61a32704

(I didn't see a need to use node.span.file.getText() but maybe you had something else in mind?)

**Comment by nickbehrens** [comment]
I looked at the crash and reduced the repro to the following:

```
@mixin loudComment {
  /* ... */
}

selector {
  @include loudComment;
}
```
In short, `@include` breaks the assumption of the current `_isTrailingComment` logic that `node.span.start.offset` will be strictly after `prevNode.span.start.offset`.

I _think_ we just need to check for `searchFrom < 0` and if so, it's never a trailing comment.  I'll commit that change (with a unit test) and see what happens with the spec tests.

**Comment by nickbehrens** [comment]
After the most recent SassSpec test run, there are 4 remaining issues.  2 of them I think are reasonable formatting changes (i.e. we could update SassSpec test cases to match):

1) [SassSpec::Test#test__spec/libsass-closed-issues/issue_894](https://github.com/sass/sass-spec/blob/master/spec/libsass-closed-issues/issue_894.hrx)
```
--- expected
+++ actual
-"a { /**/
-}
+"a { /**/ }
```

2) [SassSpec::Test#test__spec/css/keyframes/bubble/empty](https://github.com/sass/sass-spec/blob/master/spec/css/keyframes.hrx)
```
--- expected
+++ actual
-"@keyframes { /**/
-}
+"@keyframes { /**/ }
```

One of them involves a standalone loud comment following a block end (rbrace) that's now getting treated as a trailing comment.  I can add a test case for this and investigate.

3) [SassSpec::Test#test__spec/libsass-closed-issues/issue_1007](https://github.com/sass/sass-spec/blob/master/spec/libsass-closed-issues/issue_1007.hrx)
```
--- expected
+++ actual
 foo baz {
   /* before */
   margin: 0; /* after */
-}
-/* end */
+} /* end */
```

The last one I could use some help parsing/understanding:

4) [SassSpec::Test#test__spec/directives/use/css/order/import_order/comments_and_imports](https://github.com/sass/sass-spec/blob/master/spec/directives/use/css/order.hrx) 
```
--- expected
+++ actual
 "/* upstream comment before import */
-@import \"upstream.css\"; /* midstream comment before use */
+@import \"upstream.css\";
+/* midstream comment before use */
 /* midstream comment before first import */
 @import \"midstream1.css\";
 /* midstream comment before second import */
-@import \"midstream2.css\"; /* input comment before use */
+@import \"midstream2.css\";
+/* input comment before use */
 /* input comment before import */
 @import \"input.css\";
 /* upstream comment after import */
```

e.g. the expected output seems to have `@import \"upstream.css\"; /* midstream comment before use */` all on one line but I don't see that in the `.hrx` file (?)

File: test/output_test.dart [hunk 16]
```diff
@@ -162,5 +162,16 @@ selector[href*=\"{\"] { /* please don't move me */ }
 
 @rule3;"""));
     });
+
+    test("loud comments in mixin", () {
```
Comment (nex3) [inline test/output_test.dart]: This case raises a couple other possible edge cases:

* What if the loud comment comes after the style rule but still isn't contained within it, for example when the style rule is also in a mixin?
* What if the loud comment is in a totally different file from the mixin?

I'd probably solve these by writing a `bool spanContainsLocation(SourceSpan span, SourceLocation location)` helper and checking that in the outer conditional.

**Comment by Goodwine** [comment]
So taking back this conversation to what Nick spotted I saw the following problems:

1. The formatting issues spotted from before were solved. (e.g. `@keyframes { /**/ }`) 🆗 

2. Unsure what the correct format is for loud comments after open and close braces, e. g.
   ```scss
   foo { /* open */
     margin: none;
   } /* close */
   ```
   currently this becomes ... I assume this is fine?
   ```css
   foo {
     /* open */
     margin: none;
   }
   /* close */
   ```
   but there are very odd cases with nesting 😵‍💫 

3. I'm confused why the loud comments from comments and imports should become trailing comments instead of staying where they are originally 🤔

File: lib/src/visitor/serialize.dart [hunk 17]
```diff
@@ -1287,44 +1289,88 @@ class _SerializeVisitor
...
+    if (node is! CssComment) return false;
+
+    var previousSpan = previous.span;
+    if (previous is CssParentNode && previous.children.contains(node)) {
+      // Walk back from just before the current node starts looking for the
+      // parent's left brace (to open the child block).  This is safer than
+      // a simple forward search of the previousSpan.text as that might contain
+      // other left braces.
+      var searchFrom = node.span.start.offset - previousSpan.start.offset - 1;
+      if (searchFrom < 0) {
```
Comment (nex3) [inline lib/src/visitor/serialize.dart:1356]: This isn't necessarily the most reliable determiner of whether the math on the span offsets is nonsense... it's possible that `node.span` appears late in a different file than `previousSpan`, for example. It's probably better to write a little `SpanExtensions.contains()` utility and check that `previous.span.contains(node.span)`. You could even use that instead of the `previous.children.contains(node)` check.

File: test/output_test.dart [hunk 18]
```diff
@@ -110,4 +110,210 @@ void main() {
...
+selector[href*="{"]
+{ /* please don't move me */ }
+
+@rule3;"""), equals("""@rule1;
+@rule2;
+selector[href*="{"] { /* please don't move me */ }
+
+@rule3;"""));
+    });
+    group("loud comments in mixin", () {
```
Comment (nex3) [inline test/output_test.dart:216]: Nit: missing newline above

File: test/output_test.dart [hunk 19]
```diff
@@ -110,4 +110,210 @@ void main() {
...
+@mixin loudComment {
+  /* ... */
+}
+
+selector {
+  @include loudComment;
+}"""), "selector {\n  /* ... */\n}");
+      });
+
+      test("with stylerule", () {
```
Comment (nex3) [inline test/output_test.dart:228]: Nit

```suggestion
      test("with style rule", () {
```

File: lib/src/visitor/serialize.dart (hunk 17 above)
Comment (Goodwine) [reply lib/src/visitor/serialize.dart:1356]: WIP

File: lib/src/visitor/serialize.dart [hunk 20]
```diff
@@ -1287,44 +1289,88 @@ class _SerializeVisitor
...
-      return;
-    }
 
-    _writeLineFeed();
-    CssNode? previous_;
-    _indent(() {
-      for (var i = 0; i < children.length; i++) {
-        var child = children[i];
-        if (_isInvisible(child)) continue;
+    CssNode? prePrevious_;
```
[the remark this answers is not in the document]
Comment (Goodwine) [reply lib/src/visitor/serialize.dart:1295]: nvm, the previous code does have the underscore, but I removed it anyways

**Review by Goodwine — COMMENTED** [review]
one comment still WIP

File: lib/src/visitor/serialize.dart (hunk 17 above)
Comment (Goodwine) [reply lib/src/visitor/serialize.dart]: IIUC then I think this is done :)

**Review by nex3 — APPROVED** [review]
